ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/HeartbeatAgentFunctionalTest.java

```diff
@@ -0,0 +1,265 @@
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_SEND_INTERVAL_MS_CONFIG, 600000)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_CHECK_INTERVAL_MS_CONFIG, 200)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_CONFIG, 2000)
+      .withProperties(ClientTrustStore.trustStoreProps())
+      .build();
+  private static final TestKsqlRestApp REST_APP_1 = TestKsqlRestApp
+      .builder(TEST_HARNESS::kafkaBootstrapServers)
+      .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:8089")
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_ENABLE_CONFIG, true)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_SEND_INTERVAL_MS_CONFIG, 600000)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_CHECK_INTERVAL_MS_CONFIG, 200)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_CONFIG, 2000)
+      .withProperties(ClientTrustStore.trustStoreProps())
+      .build();
+
+  @ClassRule
+  public static final RuleChain CHAIN = RuleChain
+      .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
+      .around(TEST_HARNESS)
+      .around(REST_APP_0)
+      .around(REST_APP_1);
+
+  @BeforeClass
+  public static void setUpClass() {
+    TEST_HARNESS.ensureTopics(2, PAGE_VIEW_TOPIC);
+    TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER, Format.JSON);
+    RestIntegrationTestUtil.createStream(REST_APP_0, PAGE_VIEWS_PROVIDER);
+    RestIntegrationTestUtil.makeKsqlRequest(
+        REST_APP_0,
+        "CREATE STREAM S AS SELECT * FROM " + PAGE_VIEW_STREAM + ";"
+    );
+  }
+
+  @Before
+  public void setup() {
+    REST_APP_0.start();
+    REST_APP_1.start();
+  }
+
+  @After
+  public void tearDown() {
+    REST_APP_0.stop();
+    REST_APP_1.stop();
+  }
+
+  @Test(timeout = 60000)
+  public void shouldMarkServersAsUp() {
+    // Given:
+    waitForClusterToBeDiscovered();
+    waitForRemoteServerToChangeStatus(this::remoteServerIsDown);
+
+    // When:
+    sendHeartbeartsEveryIntervalForWindowLength(100, 3000);
+    final ClusterStatusResponse clusterStatusResponseUp = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsUp);
+
+    // Then:
+    assertThat(clusterStatusResponseUp.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponseUp.getClusterStatus().get(host1.toString()).getHostAlive(), is(true));
+  }
+
+  @Test(timeout = 60000)
+  public void shouldMarkRemoteServerAsDown() {
+    // Given:
+    waitForClusterToBeDiscovered();
+
+    // When:
+    ClusterStatusResponse clusterStatusResponse = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsDown);
+
+    // Then:
+    assertThat(clusterStatusResponse.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponse.getClusterStatus().get(host1.toString()).getHostAlive(), is(false));
+  }
+
+  @Test(timeout = 60000)
+  public void shouldMarkRemoteServerAsUpThenDownThenUp() {
```

Comment:
we can do this in a follow-up PR, but I'm worried that these tests will add a non-trivial amount of time to our testing suite without necessarily adding much coverage. this is, if i'm understanding correctly, trying to cover the algorithm that computes missing intervals. the `HeartbeatAgentTest` already does this. so while this test is great during development to make sure that things work, I don't know if we need it in our testing suite

also any test that involves timing is prone to flakiness - can you please run this test at least 50 times and make sure it passes 100% of the time?

Reply:
These tests don't actually check the algorithm for missing heartbeats but rather the whole heartbeating process. It uses the `/clusterStatus` endpoint and all three services to check whether servers are reported alive or dead correctly. It tests the interplay of all three services